graphql_compiler/query_pagination/pagination_planning.py

```diff
@@ -1,53 +1,203 @@
+            # value distribution and make some pagination plan with more than 1 page, but
+            # the cost estimator would need to match the behavior and assume uniform value
+            # distribution when estimating the selectivity of range filters.
+            page_capacity = 1
+        elif len(quantiles) - 1 < number_of_pages:
+            # If we have some quantiles but not enough, we generate a page for each chunk
+            # of values separated by the quantiles. N quantiles split the domain into N - 1 chunks:
+            # the 0% and 100% quantiles capture the min and max values in the domain, and the 
+            # remaining N - 2 describe the distribution in between.
+            page_capacity = len(quantiles) - 1
+
+        # If we have fewer than the ideal number of quantiles, we advise creating more.
+        if quantiles is None or len(quantiles) < ideal_quantile_resolution:
+            current_quantile_resolution = 0 if quantiles is None else len(quantiles)
+            advisories = (
+                InsufficientQuantiles(
+                    pagination_node,
+                    pagination_field,
+                    current_quantile_resolution,
+                    ideal_quantile_resolution,
+                ),
+            )
+
+    # Construct and return the plan and advisories.
+    if page_capacity <= 1:
+        return None, advisories
+    plan = VertexPartitionPlan((pagination_node,), pagination_field, page_capacity)
+    return plan, advisories
+
+
+def get_pagination_plan(
+    schema_info: QueryPlanningSchemaInfo, query_ast: DocumentNode, number_of_pages: int
+) -> Tuple[PaginationPlan, Tuple[PaginationAdvisory, ...]]:
+    """Make a best-effort PaginationPlan and advise on how to improve statistics.
 
+    Might paginate to fewer than the desired number of pages if unable to find a more satisfactory
+    plan, and may even return plans with only a single page. For example, this can happen when
+    the captured statistics are insufficient, or when the planner is not smart enough to find a
+    good plan.
 
-# TODO(bojanserafimov): Make this function return a best effort pagination plan
-#                       when a good one is not found instead of returning None.
-def try_get_pagination_plan(schema_info, query_ast, number_of_pages):
-    """Make a PaginationPlan for the given query and number of desired pages if possible."""
+    If the issue can be fixed it will return alongside the result a tuple of PaginationAdvisory
+    objects that indicate why the desired pagination was not possible. Each PaginationAdvisory
+    states the necessary step that may be taken to avoid it in the future.
+
+    If the returned plan has fewer than the desired number of pages and no advisories are returned
+    then this vertex is inherently not a good vertex to paginate on, and the caller of this
+    function should consider other vertices for pagination.
```

Comment:
@obi1kenobi slightly modified your suggested docstring to add this. Some motivation of why it's like this:

Consider the next version of the pagination planner, that considers vertices other than just the root of the query. That planner will have to examine the pagination capacity of each vertex, and maybe combine multiple vertex partition plans to meet the desired number of pages. In this scenario, we can't make the leap that "if no perfect vertex partition plan is found, some action can be taken to improve statistics". We can only say that "If the total pagination capacity of all vertex partition plans we can make are insufficient, some action can be taken to improve statistics".